notebooks/classify/signature.py

```diff
@@ -0,0 +1,249 @@
+                    cleanTable = cleanTable.drop([feature], axis=1)
+                    # print('{feature} due to being a match with regular expression: {regex}'.format(feature=feature, regex=regex))
+                    break
+
+        if removeDuplicates:
+            # Finds empty columns to prevent them from being dropped
+            emptyCol = []
+            for column in cleanTable:
+                # Convert to numpy
+                data = cleanTable[column].to_numpy() 
+                if (data[0] == np.nan or data[0] == '') and (data[0] == data).all():
+                    emptyCol.append(column)
+                    
+            # Copy columns over to be added back after duplicates are removed
+            col = cleanTable[emptyCol]
+            
+            # Transpose the cleaned table and drop duplicate rows. Re-transpose to get back to the original table
+            cleanTable = cleanTable.T.drop_duplicates().T
+            
+            # Add empty columns back into table and reorder
+            cleanTable = pd.concat([cleanTable, col], axis=1)
+
+        return cleanTable
+
+    def binarizeData(self, table: pd.DataFrame, removeNaN: bool = True):
+        """[Function for binarizing a table given the presence or absence of data. i.e. if the first two columns of a dataframe row are empty followed by two present the binarized form would be 0011]
+
+        Args:
+            table (pd.DataFrame): [Pandas dataframe to binarize]
+            removeNaN (bool, optional): [Boolean that describes whether NaN values should be removed]. Defaults to True.
+
+        Returns:
+            [pd.Dataframe]: [Returns binarized table]
+        """
+
+        # Replace empty cells with NaN 
+        if removeNaN:
+            binTable = table.replace(r'^\s*$', np.nan, regex=True)
+
+        # Replace NaN values with 0 and all others with 1
+        binTable = binTable.notnull().astype('int')
+
+        return binTable
+
+    def getPresentColumns(self, signature: str, columns: List[str]):
+        """[Function that returns a list of missing and present features given the binary data signature]
+
+        Args:
+            signature (str): [string of 0's and 1's representing the data signature]
+            columns (List[str]): [list of columns in the signature]
+
+        Returns:
+            [List[str]]: [List of present columns and missing columns]
+        """
+        present = []
+        missing = []
+        for index in range(len(signature)):
+            if int(signature[index]):
+                present.append(columns[index])
+            else:
+                missing.append(columns[index])
+        return present, missing
+
+    def countTypes(self, row: pd.Series, columns: List[str], presentFeatures: List[str], featureDict: Dict):
+        """[Function that counts the number of times a datapoint shows up in the features. For example, it counts how many times the IP 44.150.161.58 shows up in the clientIP column]
```

Comment:
Try to keep lines to ~ 80 chars
Avoid the "[ ]"
Also try to keep the naming of the function as descriptive as poss.
count_feature_unique_vals

(words like "type" and "datapoint" don't really convey much meaning")

Example (keeping to 80 chars) with numpy doc standard.
(You can use Notes and Examples for further text)

"""
Return the number of occurrences of datapoints in Features.

For example....

Parameters
------------
...

Returns
--------
...

Notes
-------